### ISSUE TITLE
AWS Auth Method - IAM works only once. Login fails after token reset. 

### ISSUE THREAD

**Issue by vperederii** [rajanadar/VaultSharp#231]
**Describe the feature request or question**
Currently AWS Auth Method - IAM implementation doesn't support token reset scenario. Let's assume that Vault issues tokens with finite duration, let's say 24h. So after 24h wen need to get another token. If we configure Vault client like stated [here](https://github.com/rajanadar/VaultSharp#aws-auth-method---iam), we'll end up with stale AWS signature, since it is valid only for 15 minutes: https://www.vaultproject.io/docs/auth/aws#comparison-of-the-iam-and-ec2-methods.

**Stack trace**
```
VaultSharp.Core.VaultApiException: {"errors":["error making upstream request: received error code 403 from STS: \u003cErrorResponse xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\"\u003e\n  \u003cError\u003e\n    \u003cType\u003eSender\u003c/Type\u003e\n    \u003cCode\u003eSignatureDoesNotMatch\u003c/Code\u003e\n    \u003cMessage\u003eSignature expired: 20211112T152532Z is now earlier than 20211115T091734Z (20211115T093234Z - 15 min.)\u003c/Message\u003e\n  \u003c/Error\u003e\n  \u003cRequestId\u003ec44b6f0f-2425-4254-aa88-c2a00f7a4c59\u003c/RequestId\u003e\n\u003c/ErrorResponse\u003e\n"]}

   at VaultSharp.Core.Polymath.MakeRequestAsync[TResponse](String resourcePath, HttpMethod httpMethod, Object requestData, IDictionary`2 headers, Boolean rawResponse, Action`1 postResponseAction)
   at VaultSharp.Core.Polymath.MakeVaultApiRequest[TResponse](String resourcePath, HttpMethod httpMethod, Object requestData, Boolean rawResponse, Action`1 postResponseAction, String wrapTimeToLive, Boolean unauthenticated)
   at VaultSharp.V1.AuthMethods.AWS.AWSAuthMethodLoginProvider.GetVaultTokenAsync()
   at VaultSharp.Core.Polymath.MakeVaultApiRequest[TResponse](String resourcePath, HttpMethod httpMethod, Object requestData, Boolean rawResponse, Action`1 postResponseAction, String wrapTimeToLive, Boolean unauthenticated)
   at VaultSharp.V1.AuthMethods.Token.TokenAuthMethodProvider.LookupSelfAsync()
```

Is it possible to add ability to generate AWS request header on every Vault `login` call? Currently this information is persisted forever inside [AWSAuthMethodLoginProvider.cs](https://github.com/rajanadar/VaultSharp/blob/master/src/VaultSharp/V1/AuthMethods/AWS/AWSAuthMethodLoginProvider.cs).

I think that this feature would be very useful.

**Comment by rajanadar** [comment]
Hi @vperederii

There is a gap somewhere. VaultSharp does a login only once, in the whole lifetime of the vault client object. This login is when you access the vault api functionality for the first time. (lazy login) You can override this, using the perform immediate login feature.

Either ways, once vault client has the vault token from the aws details, it doesn't need the aws details anymore and doesn't use it.

Can you expand on your specific problem a bit more? Can you reinitailize your vault client and auth info as well?

**Comment by vperederii** [comment]
Hi @rajanadar.

You are right. Vault does the login only once. But the problem is that tokens have finite duration and we need to reset them as recommended [here](https://github.com/rajanadar/VaultSharp#force-re-fetch-of-vault-login). I do this periodically. But at the moment I reset token AWS signature already expired.

Yes, I can re-initialize Vault client, but it seems not really good workaround, since every consumer need to write a lot of custom code to cover common scenario of token refresh. Can it be a part of VaultClient?

**Comment by rajanadar** [comment]
@vperederii  The automatic refresh of login token is deliberately out of scope of VaultSharp. This is because, it involves timers and background sorcery and error handling and statefulness etc. Not technically difficult, but is to keep VaultSharp simple.

Please let me know if any other concerns.

**Comment by rajanadar** [comment]
@vperederii

What I'll do is try to solve this use case and release a beta version for you. You please let me know, if any issues after that and any pinpoints or weird behavior.

**Comment by vperederii** [comment]
@rajanadar 

That would be a great improvement. I solved issues with automatic token refresh and AWS signature generation for my project. But these are not the best solutions. Anyways, I can share my workarounds if you want.

**Comment by rajanadar** [comment]
Yes please share.

I am thinking, even if I do a timer logic, merely logging in won't work for all Auth methods. E.g. Your case of aws would fail on signature, so the whole process needs to be redone

**Comment by vperederii** [comment]
@rajanadar 

Please check this repository: https://github.com/vperederii/VaultSharpExtensions

There are two classes: `TokenRenewableVaultClient` and `InstanceProfileAWSAuthMethodInfo`. The first one implements periodic renewal of the token the latter generates AWS signature for each /login request.

However I found it very difficult to embed these classes in to client lib code because VaultSharp lib targets to many runtimes. Is there any specific reason to support older runtime versions such as .Net Framework 4.5 and .Net Standard?

**Comment by rajanadar** [comment]
ok thanks @vperederii . I missed this message notification,, just saw this. I'll send out a beta after the weekend

**Comment by std66** [comment]
I have also encountered a very similar issue. I created my client using AWS IAM auth, and after the token expired, I started receiving 'errors: ["permission denied"]'. I'm pretty sure that resetting the client token would have worked for some time, but then the AWS-signed information would have expired as well, which would have resulted in the error message above.

So, my big brain solution was to have an abstract factory + caching component, something like this:
```csharp
interface IVaultClientFactory {
  IVaultClient GetClient();
}

class VaultClientFactory : IVaultClientFactory {
  private readonly object lockObject = new();
  private readonly Func<VaultClientSettings> settingsFactoryMethod;
  private readonly bool recreateClientInsteadOfResettingToken;
  private IVaultClient client = null;
  private DateTimeOffset authExpirationTime = DateTimeOffset.MinValue;

  public VaultClientFactory(Func<VaultClientSettings> settingsFactoryMethod, bool recreateClientInsteadOfResettingToken) {
    this.settingsFactoryMethod = settingsFactoryMethod;
    this.recreateClientInsteadOfResettingToken = recreateClientInsteadOfResettingToken;
  }

  public IVaultClient GetClient() {
    if (client == null || authExpirationTime < DateTimeOffset.UtcNow) {
      lock (lockObject) {
        ConfigureClient();
      }
    }
    
    return client;
  }

  private void ConfigureClient() {
    using JoinableTaskContext taskContext = new();
    using JoinableTaskFactory jtf = new(taskContext);

    if (client == null || recreateClientInsteadOfResettingToken) {
      this.client = new VaultClient(settingsFactoryMethod());
    }
    else if (!recreateClientInsteadOfResettingToken) {
      this.client.V1.Auth.ResetAuthToken();
    }

    authExpirationTime = DateTimeOffset.Parse(jtf.Run(async () => await this.client.V1.Auth.Token.LookupSelfAsync()).Data.ExpireTime);
  }
}
```

Of course, this way the user code has to obtain the IVaultClient through this factory each time it wants to do some Vault operation, like this:
```csharp
public async Task DoSomething(IVaultClientFactory vaultFactory) {
  JToken json = await vaultFactory.GetClient().V1.Secrets.KeyValue.V2.ReadSecretAsync<JToken>("something");
  //...
}
```

I know that IVaultClient has operations that would not require authentication, so this solution needs some more fine-tuning.

However, I just need this to implement a custom Microsoft.Extensions.Configuration config provider that polls Vault every 5 minutes for changes.

**Comment by rajanadar** [comment]
hi @vperederii , this is getting delayed to new year

**Comment by tito-villalobos** [comment]
@vperederii Thanks for that pattern. This issue was exactly the problem we had. Implemented a different version that performs AssumeRoleWithWebIdentityRequest.
@rajanadar thanks for such a clean class structure making these kinds of extensions straightforward.

**Comment by rajanadar** [comment]
that is great to hear @tito-villalobos . cheers.

@vperederii , confirming that you would still find a beta release valuable? let me know, and i'll publish

**Comment by vperederii** [comment]
@rajanadar yes, a beta release with new functionality would be really helpful.

**Comment by rajanadar** [comment]
hi @vperederii 

I tried adding the capabilities into VaultSharp for a beta release and it doesn't build because the AWS and memory cache dependencies don't work across all .NET frameworks.

Please continue using the extensions you built. I'll close this for now.
